### PR TITLE
#721: Add matcher for TeeInput class

### PR DESCRIPTION
**Pull request by proshin-roman**
It is for issue #721: adds a new matcher that checks result of TeeInput class: compares both TeeInput's results (direct and copied) to the expected value. As well this new matcher has been applied to existing tests of TeeInput class.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#731](https://github.com/yegor256/cactoos/issues/731) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#731](https://github.com/yegor256/cactoos/issues/731) is assigned to @neonailol[/z](https://www.0crat.com/u/neonailol), here is [why](https://www.0crat.com/footprint/C63314D6Z/691536760). The budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4). Please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

**Comment by proshin-roman** [comment]
@neonailol ping :)

File: src/main/java/org/cactoos/matchers/TeeInputHasResult.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+    /**
+     * Ctor.
+     * @param expected Value that is expected to be returned from the TeeInput
+     * @param copied Value that is expected to be copied into the output of
+     *  the TeeInput
+     */
+    public TeeInputHasResult(
+        final Text expected,
+        final Text copied) {
```
Comment (neonailol) [inline src/main/java/org/cactoos/matchers/TeeInputHasResult.java]: formatting is not consistent

File: src/main/java/org/cactoos/matchers/TeeInputHasResult.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+    /**
+     * Ctor.
+     * @param expected Value that is expected to be returned from the TeeInput
+     * @param copied Value that is expected to be copied into the output of
+     *  the TeeInput
+     */
+    public TeeInputHasResult(
+        final String expected,
+        final Text copied) {
```
Comment (neonailol) [inline src/main/java/org/cactoos/matchers/TeeInputHasResult.java]: formatting is not consistent

File: src/main/java/org/cactoos/matchers/TeeInputHasResult.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        this.expected = new ComparableText(expected);
+        this.copied = new ComparableText(copied);
+        this.actual = new ComparableText(new TextOf(""));
+    }
+
+    @Override
+    public boolean matchesSafely(final TeeInput item) {
+        this.actual = new ComparableText(new TextOf(item));
+        return
+            this.expected.compareTo(this.actual) == 0
```
Comment (neonailol) [inline src/main/java/org/cactoos/matchers/TeeInputHasResult.java]: Please use `org.cactoos.scalar.Equals` here

File: src/main/java/org/cactoos/matchers/TeeInputHasResult.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    @Override
+    public void describeTo(final Description description) {
+        description.appendText(
+            new UncheckedText(
+                new FormattedText(
+                    TeeInputHasResult.DESCRIPTION,
+                    new UncheckedText(this.expected).asString(),
+                    new UncheckedText(this.expected).asString()
```
Comment (neonailol) [inline src/main/java/org/cactoos/matchers/TeeInputHasResult.java]: Duplicated line `new UncheckedText(this.expected).asString()`

**Comment by proshin-roman** [comment]
@neonailol Thanks for your comments - they were useful. I've fixed all of them except a comment about inconsistent formatting: I've changed another ctor where the closing bracket `)` is on a new line. Please review the pull request again.

**Comment by neonailol** [comment]
@proshin-roman thanks looks good now

**Comment by neonailol** [comment]
@rultor good to merge

**Comment by proshin-roman** [comment]
@yegor256 ping

**Comment by neonailol** [comment]
@yegor256 Please confirm this.

**Comment by yegor256** [comment]
@proshin-roman to be honest, I don't like this at all. Why do we need this matcher? What's wrong with existing `assertThat` two times? It looks like something very artificial and super-specific for one particular use case.

**Comment by proshin-roman** [comment]
@yegor256 well, TeeInput class has a lot of ctors (about 50) and all of them should be covered by tests (there is a ticket for that). It means double-assert will be copy-pasted for all these test cases. Another problem of double-assert: you have to trigger them in exact order otherwise they fail. This new matcher allows to avoid both of these problems. Finally, it's a goal of the ticket - you can decline this feature at all as the Arch of the project. And then I will have a reason to use double-assert in my other pull request.

**Comment by 0crat** [comment]
@neonailol[/z](https://www.0crat.com/u/neonailol) this job was assigned to you 8 days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by yegor256** [comment]
@rultor merge

**Comment by proshin-roman** [comment]
@yegor256 damn, the pull request is going to be merged but the ticket is already resigned from me. I would be very happy if you assign it back to me, please :)

**Comment by 0crat** [comment]
@elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova) please review this job completed by @neonailol[/z](https://www.0crat.com/u/neonailol), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#731](https://github.com/yegor256/cactoos/issues/731) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256)

**Comment by elenavolokhova** [comment]
@neonailol According to our Policy:
>The code reviewer found at least three problems in the code

Only two issues were fixed and actually improved the code quality in this case. 
Please, confirm that you will try to find at least three problems next time.

**Comment by neonailol** [comment]
@elenavolokhova i confirm

**Comment by elenavolokhova** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 points just awarded to @neonailol[/z](https://www.0crat.com/u/neonailol)

**Comment by 0crat** [comment]
Quality review completed: +8 points just awarded to @elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova)

**Pull request merged**